### DISCUSSION TITLE
Export an airplane as STEP or, if possible, STL

### DISCUSSION THREAD

**Discussion by chiefenne** [peterdsharpe/AeroSandbox#106] [Q&A] [answered]
I would like to export the airplane in a format like STEP or STL.
I found in airplane.py that export_cadquery_geometry does the job to export STEP which is really cool.

But I can't get it to work. It complains about a wrong numpy version, I guess of the cadquery API.

**Comment by peterdsharpe** [comment, marked as answer]
Hmmm, interesting!

For the STL case, an immediate workaround is to instead bypass CadQuery and use the alternate PyVista API:
```
plotter = airplane.draw(backend="pyvista", show=False)
plotter.save("airplane.stl")
```

However, in general I think STEP files are more useful than STLs for supporting ongoing engineering, so I'd like to figure out what's causing the bug you're seeing with STEP export.

Can you send a self-contained minimum reproducible example of this bug (both code and traceback) along with your versions of AeroSandbox, cadquery, NumPy, and Python?

Thank you!

**Comment by chiefenne** [comment]
Hi,
when I tried to export the STEP I did this on a Windows Laptop. Meanwhile I tried the export also on a MAC (with a fresh environment). Both, the STEP and STL exports worked flawlessly.
I think I need to start over on the Windows machine with a fresh environment, so that we do not chase a phantom. I can do this tomorrow. If I fail in the fresh environment I will compile a self-contained example.

On the MAC the following versions are installed:
```
Python version: 3.11.4
AeroSandbox version: 4.1.1
AeroSandbox-NumPy version: 1.24.4
SciPy version: 1.11.2
Matplotlib version: 3.7.2
CasADi version: 3.6.3
CadQuery version: 2.3.1
```

Thanks for your support and congrats to this outstanding piece of software.

**Comment by chiefenne** [comment]
Hi,

meanwhile I created a fresh environment on Windows and it also worked without any problems.
Sorry for any inconvenience I caused.

I noticed one thing though, The "pip install" did not install cadquery, although available in requirements. I couldn't find out why.

```
Python version: 3.11.5 | packaged by conda-forge | (main, Aug 27 2023, 03:23:48) [MSC v.1936 64 bit (AMD64)]
AeroSandbox version: 4.1.1
AeroSandbox-NumPy version: 1.24.4
SciPy version: 1.11.2
Matplotlib version: 3.7.3
CasADi version: 3.6.3
Cadquery version: 2.3.1
```

**Reply by peterdsharpe** [reply]
No worries, I'm glad this worked!

`pip install aerosandbox` does not install cadquery, since cadquery is categorized as an "optional dependency" of AeroSandbox.

So:

* `pip install aerosandbox` will install the "minimum-viable set of dependencies needed to work with aerosandbox" - basically just numerics and optimization modules, with limited visualization or I/O support.
* `pip install aerosandbox[full]` will install all of the optional features, bells, and whistles (including cadquery)

More info here: https://github.com/peterdsharpe/AeroSandbox/blob/40e22c4148a1c53681dfeab19dd4b252ca6b228d/setup.py#L67

The motivation for this was that some people were using AeroSandbox in embedded control systems (basically doing model-predictive control) and needed to install AeroSandbox on unconventional hardware - so this minimized dependency conflicts.

**Comment by peterdsharpe** [comment]
Just adding onto this discussion (for future readers) that `asb.Airplane.export_OpenVSP_vspscript()` is another means to export AeroSandbox designs to CAD. This will produce a vspscript file that can be read by [OpenVSP](openvsp.org), and from there it can be tweaked and exported to your favorite 3D CAD tool.
